### PR TITLE
Add invariant checking

### PR DESCRIPTION
**Pull request by sjakobi**
Addresses #366.

---

TODO:

* [ ] Add more tests
* [x] Document invariants
* [x] Extract validation code into `D.HM.I.Debug`?!

### CODE REVIEW

**Comment by sjakobi** [comment]
I've checked that this tooling could have caught #420. Sample output:

```
        intersection produces valid HashMap: FAIL
          *** Failed! Falsified (after 61 tests and 10 shrinks):
          fromList [(K {hash = 265, _x = D},()),(K {hash = 16649, _x = A},())]
          fromList [(K {hash = 265, _x = D},()),(K {hash = 16649, _x = A},())]
          Invalid (INV2_misplaced_hash 265) (Cons 0 Root) /= Valid
          Use --quickcheck-replay=665566 to reproduce.
          Use -p '/intersection/&&/Data.HashMap.Lazy.difference and intersection.intersection produces valid HashMap/' to rerun this test only.
```

**Comment by sjakobi** [comment]
Hmm, is there another bug in the new intersection implementation?! oO

**Comment by sjakobi** [comment]
> Hmm, is there another bug in the new intersection implementation?! oO

Probably not. I'm getting the same errors with the naive old `intersection` implementation. I must have made a mistake in the validation logic.

File: Data/HashMap/Internal.hs [hunk 1]
```diff
@@ -2456,3 +2465,90 @@ instance (Eq k, Hashable k) => Exts.IsList (HashMap k v) where
...
+    validInternal p (Leaf h l)            = validHash p h <> validLeaf p h l
+    validInternal p (Collision h ary)     = validHash p h <> validCollision p h ary
+    validInternal p (BitmapIndexed b ary) = validBitmapIndexed p b ary
+    validInternal p (Full ary)            = validFull p ary
+
+    validHash p0 h0 = go p0 h0
+      where
+        go Root !_ = Valid
+        go (Cons sh p) h | h .&. subkeyMask == sh = go p (h `unsafeShiftR` bitsPerSubkey)
+                         | otherwise              = Invalid (INV2_misplaced_hash h0) p0
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:2515]: I suspect the bug is in here. We seem to match up the hash with the sub-hash path in inverted order.

**Comment by sjakobi** [comment]
This should be ready for review. I'll hold off on adding more validity tests until I've refactored the existing property tests a bit.

File: Data/HashMap/Internal/Debug.hs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+data Error k
+  = INV1_internal_Empty
+  | INV2_bad_BitmapIndexed_size !Int
+  | INV3_bitmap_array_size_mismatch !Bitmap !Int
+  | INV4_BitmapIndexed_invalid_single_subtree
+  | INV5_misplaced_hash !Hash
+  | INV6_key_hash_mismatch k !Hash
+  | INV7_bad_Full_size !Int
+  | INV8_Collision_size !Int
+  | INV9_Collision_duplicate_key k !Hash
```
Comment (treeowl) [inline Data/HashMap/Internal/Debug.hs:68]: What are the invariant numbers for?

File: Data/HashMap/Internal/Debug.hs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    validInternal p (Leaf h l)            = validHash p h <> validLeaf p h l
+    validInternal p (Collision h ary)     = validHash p h <> validCollision p h ary
+    validInternal p (BitmapIndexed b ary) = validBitmapIndexed p b ary
+    validInternal p (Full ary)            = validFull p ary
+
+    validHash p0 h0 = go (reverse p0) h0
+      where
+        go [] !_ = Valid
+        go (sh:p) h | h .&. subkeyMask == sh = go p (h `unsafeShiftR` bitsPerSubkey)
+                    | otherwise              = Invalid (INV5_misplaced_hash h0) p0
```
Comment (treeowl) [inline Data/HashMap/Internal/Debug.hs:92]: I'm not wild about doing O(depth) calculation for every hash in every validity test. Can't we recursively pass down the low bits we expect, and a mask for how many there are, reducing that to O(1) per hash?

File: Data/HashMap/Internal/Debug.hs (hunk 2 above)
Comment (sjakobi) [reply Data/HashMap/Internal/Debug.hs:68]: Actually I'm not sure whether these will be useful. They are just reference numbers that can hopefully stay constant when we reword the invariants or rename the error constructors.

File: Data/HashMap/Internal/Debug.hs (hunk 3 above)
Comment (sjakobi) [reply Data/HashMap/Internal/Debug.hs:92]: Good point. Will do.

File: Data/HashMap/Internal/Debug.hs [hunk 4]
```diff
@@ -73,23 +71,40 @@ data Error k
...
+
+addSubHash :: SubHashPath -> SubHash -> SubHashPath
+addSubHash (SubHashPath ph l) sh =
+  SubHashPath (ph .|. (sh `unsafeShiftL` l)) (l + bitsPerSubkey)
+
+hashMatchesSubHashPath :: SubHashPath -> Hash -> Bool
+hashMatchesSubHashPath (SubHashPath ph l) h = maskToLength h l == ph
+  where
+    -- Note: This needs to use `shiftL` instead of `unsafeShiftL` because
+    -- @l'@ may be greater than 32/64 at the deepest level.
```
Comment (treeowl) [inline Data/HashMap/Internal/Debug.hs:93]: Greater than, or just equal to?

Comment (sjakobi) [reply Data/HashMap/Internal/Debug.hs:93]: `l` is 65 when the deepest leaves are checked on a 64-bit computer.

File: Data/HashMap/Internal.hs [hunk 5]
```diff
@@ -210,10 +213,40 @@ instance NFData2 Leaf where
...
+    -- * 'Empty' is not a valid sub-node. It can only appear at the root. (INV1)
     | BitmapIndexed !Bitmap !(A.Array (HashMap k v))
+    -- ^ Invariants:
+    --
+    -- * The array of a 'BitmapIndexed' node stores at least 1 and at most
+    --   @'maxChildren' - 1@ sub-nodes. (INV2)
+    -- * The number of sub-nodes is equal to the number of 1-bits in its
+    --   'Bitmap'. (INV3)
+    -- * If a 'BitmapIndexed' node has only one sub-node, this sub-node must
+    --   be a 'BitmapIndexed' or a 'Full' node. (INV4)
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:227]: There's another invariant that I should add: The upper `finiteBitSize @Word 0 - maxChildren - 1` bits of the `Bitmap` are all `0`.

Comment (sjakobi) [reply Data/HashMap/Internal.hs:227]: Correction: Only the `finiteBitSize @Word 0 - maxChildren` upper bits must be `0`. The bit at index 31 may be `1`.

**Comment by treeowl** [comment]
Huzzah!

**Pull request merged**